src/hotspot/share/gc/g1/g1CollectedHeap.hpp

```diff
@@ -871,15 +872,16 @@ class G1CollectedHeap : public CollectedHeap {
   // The parallel task queues
   G1ScannerTasksQueueSet *_task_queues;
 
-  // Number of regions evacuation failed in the current collection.
-  volatile uint _num_regions_failed_evacuation;
-  // Records for every region on the heap whether evacuation failed for it.
-  CHeapBitMap _regions_failed_evacuation;
+  G1EvacuationFailureRegions _evac_failure_regions;
 
   // Preserve the mark of "obj", if necessary, in preparation for its mark
   // word being overwritten with a self-forwarding-pointer.
   void preserve_mark_during_evac_failure(uint worker_id, oop obj, markWord m);
 
+  void iterate_evacuation_failure_regions_par(HeapRegionClosure* closure,
```

Comment:
Please avoid additional clutter of `G1CollectedHeap`. Pass `_evac_failure_regions` to `post_evacuate_collection_set` and then down to `G1ParRemoveSelfForwardPtrsTask` where it calls `par_iterate` directly.
I am currently moving stuff, particularly data structures only used in the young collector to a separate class (JDK-8253343). So while it is not avoidable to put `_evac_failure_regions` here, let's avoid further adding young-collector only stuff to `G1CollectedHeap`.

I'll of course do those changes (i.e. moving code to `G1YoungCollector`) and further related cleanup when ready.